### PR TITLE
Ampers: Kaitlin Forsman

### PR DESCRIPTION
**Pull request by kcforsman**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | I chose to return nil if there were no available drivers rather than an error. I did this because thinking about these types of app you probably wouldn't want the program to break for the passenger if no driver were available. (although you could rescue it and display a message instead)
Describe any examples of composition that you encountered in this project, if any | There was a lot of composition in this project. Every trip_dispatcher has-an Array of RideShare::Drivers, has-an Array of Rideshare::Passengers, and has-an Array of Rideshare::Trips. And other 3 classes have at least one of other classes in a compositional relationship as well. It felt very convoluted to sort through the overlap 'has-a' relationships.
Describe a _nominal test_ that you wrote for this assignment. | When writing the nominal tests for 'request_trip', I included a test that checked the return of the method was in fact an instance of a RideShare::Trip.
Describe an _edge case test_ that you wrote for this assignment. | For the case when no drivers were available, I wrote a test based on the length of available driver. It called request_trip until there were no available drivers left then tested the next call of request_trip returned nil.
Describe a concept that you gained more clarity on as you worked on this assignment. | I think I still need to work on them but I feel like I learned a lot about navigating into compositional relationships using enumerables, particularly by experimenting with wave 3.

### CODE REVIEW

File: lib/trip_dispatcher.rb [hunk 1]
```diff
@@ -90,6 +93,58 @@ def load_trips
       trips
     end
 
+    # addition that takes passenger_id to create new trip
+    def request_trip(passenger_id)
+      check_id(passenger_id)
+      id = trips.length + 1
```
Comment (CheezItMan) [inline lib/trip_dispatcher.rb:99]: For this exercise this works fine, but if the list of trips could have deleted entries, then setting the id this way could lead to duplicate ids.  

I would suggest instead:
```ruby
id = trips.max_by { |trip| trip.id }.id + 1
```

File: lib/trip_dispatcher.rb [hunk 2]
```diff
@@ -90,6 +93,58 @@ def load_trips
...
+    # if no driver is available returns nil
+    def find_least_recently_active_driver
+      recent_rides = []
+      @drivers.find_all do |driver|
+        if driver.status == :AVAILABLE && driver.trips.empty?
+          return driver
+        elsif driver.status == :AVAILABLE
+          available_drivers = []
+          unsorted_rides = []
+          available_drivers << driver
```
Comment (CheezItMan) [inline lib/trip_dispatcher.rb:132]: look at `available_drivers` you're setting it to an empty array and then inserting 1 driver.

File: lib/trip_dispatcher.rb [hunk 3]
```diff
@@ -90,6 +93,58 @@ def load_trips
...
+          available_drivers.each { |available_driver| unsorted_rides << available_driver.trips}
+          unsorted_rides.each do |rides|
+              recent_rides << rides.max_by {|ride| ride.end_time.to_i}
+          end
+        end
+      end
+      return if recent_rides.empty?
```
Comment (CheezItMan) [inline lib/trip_dispatcher.rb:139]: I would be more specific and say
```ruby
return nil if recent_rides.empty?
```

File: specs/driver_spec.rb [hunk 4]
```diff
@@ -77,4 +82,64 @@
       driver.average_rating.must_equal 0
     end
   end
+
+  describe 'total_revenue' do
```
Comment (CheezItMan) [inline specs/driver_spec.rb:86]: For `total_revenue` also test when the driver has no trips.

File: specs/driver_spec.rb [hunk 5]
```diff
@@ -77,4 +82,64 @@
...
+      @driver.add_trip(@trip)
+      @driver.add_trip(@trip2)
+      @driver.add_trip(@trip3)
+
+      @driver.total_revenue.must_be_kind_of Float
+      @driver.total_revenue.must_equal 8.04
+    end
+  end
+
+  describe 'average_hourly_revenue' do
```
Comment (CheezItMan) [inline specs/driver_spec.rb:107]: ditto

File: specs/passenger_spec.rb [hunk 6]
```diff
@@ -75,4 +79,36 @@
       end
     end
   end
+
+  describe 'total_money_spent' do
```
Comment (CheezItMan) [inline specs/passenger_spec.rb:83]: Similarly test when the passenger has no trips.

File: specs/passenger_spec.rb [hunk 7]
```diff
@@ -75,4 +79,36 @@
...
+      @passenger.add_trip(RideShare::Trip.new({id: 8, driver: nil, passenger: @passenger, start_time: start_time, end_time: end_time, rating: 5, cost: 8.00}))
+      @passenger.add_trip(RideShare::Trip.new({id: 9, driver: nil, passenger: @passenger, start_time: start_time, end_time: end_time, rating: 5, cost: 8.00}))
+      @passenger.add_trip(RideShare::Trip.new({id: 10, driver: nil, passenger: @passenger, start_time: start_time, end_time: end_time, rating: 5, cost: 8.00}))
+
+      @passenger.total_money_spent.must_be_kind_of Float
+      @passenger.total_money_spent.must_equal 24
+    end
+  end
+
+  describe 'total_time_spent' do
```
Comment (CheezItMan) [inline specs/passenger_spec.rb:99]: ditto

**Comment by CheezItMan** [comment]
# Ride Share
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | Good number of commits and very good commit messages
Answer comprehension questions | Check 
**Wave 1** |
`Trip` has a helper method to calculate duration | Check
`Passenger` has a method to calculate total cost of all trips | Check
`Driver` has a method to calculate total revenue | Check
Complete tests | See my notes
**Wave 2** |
`TripDispatcher` has a new method to create trips | Check
Creating a trip in `TripDispatcher` relies on methods in `Driver` and `Passenger` to modify their own attributes | Check
Methods from wave 1 work on `Driver`s and `Passenger`s with incomplete trips | Check
Complete tests | Check, but see my notes
**Wave 3** (optional) |
Complex logic was correctly implemented and tested | You've got some tangled and unnecessary code, but it is working.
Appropriate use of Ruby's `Date` |  Check
Appropriate helper methods were made to help with complex logic | Some yes
Complete tests | See my notes, but pretty good tests
**Overall** | Well done overall, you do need to look a bit more carefully for edge-cases and some of your wave 3 code is a bit tangled and needs refactoring.  However you definitely hit all the edge cases and did a great job on this project.
